### ISSUE TITLE
Null check incorrectly (?) optimized out for nested struct access

### ISSUE THREAD

**Issue by jkotas** [dotnet/coreclr#23791]
Repro:

```
using System;
using System.Runtime.CompilerServices;

unsafe struct F
{
    long X;
    fixed byte A[10];

    [MethodImpl(MethodImplOptions.NoInlining)]
    byte NextElement(int i) => A[1+i];

    static int Main() {
        F* x = null;
        return x->NextElement(100000);
    }
}
```

Expected result: NullReferenceException
Actual result: AccessViolationException, or passed

**Comment by jkotas** [comment]
Is it a bug or by design that the null check is optimized out in this case?

**Comment by jkoritzinsky** [comment]
What is thee supposed to be a null check around? The fixed buffer field is effectively a struct. I don't see anything that could be null here.

**Comment by jkotas** [comment]
The first IL instruction in NextElement is `ldflda`. `ldflda` is usually expected to throw NullReferenceException when it is used on null pointer.

**Comment by AndyAyersMS** [comment]
Not sure either. Will investigate.

**Comment by AndyAyersMS** [comment]
Current jit strategy is that explicit null checks are added for field (ldfld*) accesses only when the cumulative offset is "large" or unknown and the byref is either the ultimate result of the computation or feeds an indir (as it does here). 

But since the final bits of address formation that push us into "large" offset territory are regular adds and not field offsets, this logic does not kick in.

Seems like we should look into adding this checking.

Also seems odd that there is no bounds checking on the fixed field access index, which would help prevent "small" out of bounds accesses. For instance, this compiles and runs and returns 255 by indexing off the end:
```C#
using System;
using System.Runtime.CompilerServices;

unsafe struct F
{
    fixed byte A[16];

    static int Main() {

        F y = new F();
        F x = new F();
        
        for (int i = 0; i < 16; i++)
        {
            x.A[i] = 0x64;
            y.A[i] = 0xFF;
        }

        byte result = x.A[20];

        return (int) result;
    }
}
```
and likewise, one can index below the start:
```C#
using System;
using System.Runtime.CompilerServices;

unsafe struct F
{
    fixed byte A[16];

    static int Main() {

        F x = new F();
        F y = new F();
        
        for (int i = 0; i < 16; i++)
        {
            x.A[i] = 0x64;
            y.A[i] = 0xFF;
        }

        byte result = x.A[-5];

        return (int) result;
    }
}
```

cc @jaredpar

**Comment by AndyAyersMS** [comment]
Also cc @VSadov -- seems like if we're not going to bounds check the index I wonder if null checking the base adds a lot of value.

**Comment by VSadov** [comment]
There is no range check since `x.A[i]`  it is treated as a pointer indirection. Same as `*(x.A + i)`.
It may feel like compiler knows the size of A, but A is just a pointer.

There is no such thing as "fixed size array" _type_ in the language.

I.E.

if   `v = x.A[i]`   would range check, then you would expect that the following would as well.

```cs
var temp = x.A;
. . . 
v = temp[i];   
```

but it can not. The type of temp is `byte*`

**Comment by VSadov** [comment]
Considering that

```cs
        F* n = null;
        var v = n->A[offset];
```
is the same as 

```cs
        F* n = null;
        var temp = n->A;                 //  would you null-check  here?
        var v1 = *(temp + offset);
```

I would be ok with not checking the receiver of struct ldflda and perhaps treating it as not sideeffecting.

**Comment by jkotas** [comment]
> no bounds checking

It is by design that `fixed` buffers do not have bounds-checking.

**Comment by jkotas** [comment]
The non-intuitive behavior that made me file this issue is that whether or not ldflda throws NullReferenceException for null "this" depends on the instructions that follow it. The IL for the NextElement method from the C# repro at the top is:

```
ldarg.0
ldflda     valuetype F/'<A>e__FixedBuffer' F::A
ldflda     uint8 F/'<A>e__FixedBuffer'::FixedElementField
ldc.i4.1
ldarg.1
add
add
ldind.u1
ret
```

Changing IL of NextElement method to the following will make it throw NullReferenceException:

```
ldarg.0
ldflda     valuetype F/'<A>e__FixedBuffer' F::A

pop
ldc.i4 0
ret
```

But adding one more ldflda like this will make the NullReferenceException disapper:

```
ldarg.0
ldflda     valuetype F/'<A>e__FixedBuffer' F::A
ldflda     uint8 F/'<A>e__FixedBuffer'::FixedElementField
pop
ldc.i4 0
ret
```

**Comment by AndyAyersMS** [comment]
Yes, that seems odd. Let me debug what morph does with the `MorphAddrContext` in those cases.

**Comment by AndyAyersMS** [comment]
Looks like morph is losing track of context in ADDR/FIELD/ADDR/FIELD chains (as in the "one more ldflda case" and in the initial example). `fgMorphField` will modify the FIELD into an IND, and this causes morph to reset the address context as it thinks the IND will actually do a deref. But that's not true if the parent of the IND is an ADDR, as morph will later change ADDR(IND(X)) to X.

cc @dotnet/jit-contrib

**Comment by mikedn** [comment]
Oh boy, morph address context. That thing should probably be replaced by `LocalAddressVisitor`. Though that was primarily intended to deal with local addresses... but it has a more sensible approach regarding field chain morphing.

**Comment by AndyAyersMS** [comment]
Contextual child nodes certainly feel like an anti-pattern.

I think there may be a surgical fix for this particular issue. Let me poke at it a bit more. But I agree we should rethink this whole area.

**Comment by AndyAyersMS** [comment]
Have put up a proposed fix in #23850 -- it uses GTF_DONT_CSE to catch cases where IND and FIELD are part of an ongoing address chain.

**Comment by AndyAyersMS** [comment]
@jkotas I don't get exceptions for those two truncated cases with the current jit...

**Comment by jkotas** [comment]
@AndyAyersMS Do you have the `long X;` field in your repro? This field makes a difference - if you omit the field, the NullReference exception is never thrown. (I did not have it include in the minimal repro at the top originally.)

**Comment by AndyAyersMS** [comment]
Thanks, I can repro. The lack of any exception from the second case is looks like a different issue and is still there with my fix -- morph is discarding an unused tree with a potential side effect. I'll look into it.

**Comment by AndyAyersMS** [comment]
Hmm, looks like I was confused. The proposed fix in #23850 handles the second case.
